### PR TITLE
Suggest using raw identifiers in 2018 edition when using keywords

### PR DESCRIPTION
**Pull request by estebank**

### CODE REVIEW

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -797,7 +797,18 @@ impl<'a> Parser<'a> {
         let mut err = self.struct_span_err(self.span,
                                            &format!("expected identifier, found {}",
                                                     self.this_token_descr()));
-        if let Some(token_descr) = self.token_descr() {
+        if let (true, token::Ident(ref s, false), true) = (
+            self.span.rust_2018(),
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Raw identifiers are available on Rust 2015 as well.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -797,7 +797,18 @@ impl<'a> Parser<'a> {
...
+            &self.token,
+            self.token.is_used_keyword() || self.token.is_unused_keyword(),
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: The right condition here is `ident.is_reserved() && !ident.is_path_segment_keyword() && ident.name != keywords::Underscore.name()` because path segment keywords like `self` or `super` cannot be raw.

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -797,7 +797,18 @@ impl<'a> Parser<'a> {
         let mut err = self.struct_span_err(self.span,
                                            &format!("expected identifier, found {}",
                                                     self.this_token_descr()));
-        if let Some(token_descr) = self.token_descr() {
+        if let (true, token::Ident(ref s, false), true) = (
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Nit: `ref s` -> `ident`.

File: src/libsyntax/parse/parser.rs [hunk 4]
```diff
@@ -797,7 +797,18 @@ impl<'a> Parser<'a> {
...
+            self.span.rust_2018(),
+            &self.token,
+            self.token.is_used_keyword() || self.token.is_unused_keyword(),
+        ) {
+            err.span_suggestion_with_applicability(
+                self.span,
+                "you can escape reserved keywords to use them as identifiers",
+                format!("r#{}", s.to_string()),
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Nit: `Ident`s are already printable without `to_string` conversion.

File: src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr [hunk 5]
```diff
@@ -2,9 +2,13 @@ error: expected identifier, found reserved keyword `async`
   --> $DIR/edition-keywords-2015-2018-expansion.rs:8:5
    |
 LL |     produces_async! {} //~ ERROR expected identifier, found reserved keyword
-   |     ^^^^^^^^^^^^^^^^^^ expected identifier, found reserved keyword
+   |     ^^^^^^^^^^^^^^^^^^
```
Comment (petrochenkov) [inline src/test/ui/editions/edition-keywords-2015-2018-expansion.stderr]: I'd personally keep the label, it feels orthogonal to the help message.

**Comment by petrochenkov** [comment]
@estebank 
After some thought, maybe raw identifiers need to be recommended only for edition-specific keywords?
IMO, raw identifiers are more an emergency feature, than something that should be used casually.

**Comment by estebank** [comment]
@petrochenkov that sounds reasonable to me. I want to give people a chance to find out about the feature, but not to overuse it... Maybe not even using a suggestion but just linking to the book could be a good alternative. Higher barrier to finding than just accepting a change, while also having a chance to find out about them. The drawback is that I have seen people (particularly medium-experienced devs) ignoring the links even when they'd include a bit of info they wouldn't know. Maybe just an old school `help` message informing on raw idents?
